syntax = "proto3";

import "actionizers.proto";
import "events.proto";
import "featurizers.proto";

import "scalapb/scalapb.proto";

option (scalapb.options) = {
  scope: FILE
  preserve_unknown_fields: false
  package_name: "co.firstorderlabs.common.protos"
};

enum InfoDictKey {
    buyFeesPaid = 0;
    buyOrdersPlaced = 1;
    buyVolumeTraded = 2;
    numOrdersRejected = 3;
    portfolioValue = 4;
    sellFeesPaid = 5;
    sellOrdersPlaced = 6;
    sellVolumeTraded = 7;
    simulationStepDuration = 8;
    dataGetDuration = 9;
    matchingEngineDuration = 10;
    environmentDuration = 11;
    numEvents = 12;
    eventsPerMilliSecond = 13;
    numSamples = 14;
    numEpisodes = 15;
    roi = 16;
    midPrice = 17;
}

enum RewardStrategy {
    ReturnRewardStrategy = 0;
    LogReturnRewardStrategy = 1;
}

service EnvironmentService {
    rpc executeAction (ActionRequest) returns (Order) {
    }
    rpc getInfo (SimulationId) returns (Info) {
    }
    rpc getObservation (ObservationRequest) returns (Observation) {
    }
    rpc getReward (RewardRequest) returns (Reward) {
    }
}

message ActionRequest {
    repeated double actorOutput = 1;
    Actionizer actionizer = 2;
    SimulationId simulationId = 3;
    bool updateOnly = 4;
}

message Info {
    InfoDictProto infoDict = 1;
    ActionizerState actionizerState = 2;
}

message InfoDictProto {
    option (scalapb.message).type = "co.firstorderlabs.common.protos.environment.InfoDict";
    map<string, double> infoDict = 1;
    map<string, double> actionizerState = 2;
}

message ObservationRequest {
    bool normalize = 2;
    RewardRequest rewardRequest = 3;
    SimulationId simulationId = 4;
    Featurizer featurizer = 5;
    FeaturizerConfigs featurizerConfigs = 6;
}

message Feature {
    repeated double feature = 1;
}

message Features {
    map<string, Feature> features = 1;
}

message Observation {
    Features features = 1;
    Reward reward = 2;
    Info info = 3;
}

message RewardRequest {
    RewardStrategy rewardStrategy = 1;
    SimulationId simulationId = 2;
}

message Reward {
    double reward = 1;
}